Basic character statistics

mbstraus opened the pull request:
This adds simple stats to the characters and the character inspector pane.  There are 6 stats total that have been added, although these can be easily tweaked via the stats.xml file.  The stats are not actually used for anything at this point, they are purely just numbers tagged to the character.  The stats will also be saved and loaded.

Related to issue #1035 

![image](https://cloud.githubusercontent.com/assets/434820/18111702/f13ec692-6ede-11e6-88bd-9fd71065361b.png)

Apologies if I did something wrong with setting up this pull request... first time.

TomMalbran commented:
This looks good. One thing I would change is the make the Stats array a dictionary, so that in the future you can do something like ```Character.GetStat("Strength")```. Since we don't know where this will be used I am not sure about which methods we need for them, and what information we actually need (Do we need to know the min and max or just the value. Can we change them, maybe using an xp approach).

And one other question: Do we need to set the min and max? Or maybe we can standardize them to 1 and 20, so that all the skills look the same.

mbstraus commented:
@TomMalbran 

Great suggestions!  I updated it to a dictionary and added a GetStat method (old Java habits and method names starting with lower case >.<).  I also removed the min and max values... probably not needed as it stands, and is easy enough to add back in if necessary.

Grenkin1988 commented on Assets/Scripts/Models/Stat.cs:
```diff
@@ -0,0 +1,58 @@
+        this.Name = other.Name;
+    }
+
+    public void ReadXmlPrototype(XmlReader reader_parent)
```
parentReader

Grenkin1988 commented on Assets/Scripts/Models/Stat.cs:
```diff
@@ -0,0 +1,58 @@
+        statType = reader_parent.GetAttribute("statType");
+
+        XmlReader reader = reader_parent.ReadSubtree();
+        List<string> luaActions = new List<string>();
```
`luaActions` where is it used?

Grenkin1988 commented on Assets/Scripts/Models/Stat.cs:
```diff
@@ -0,0 +1,58 @@
+    public string Name;
+    public int MinValue;
+    public int MaxValue;
+    public Character character;
```
Why `Stat` should know about `Character`?

Grenkin1988 commented on Assets/StreamingAssets/Data/Stats.xml:
```diff
@@ -0,0 +1,21 @@
+<Stats>
+  <Stat statType="Strength">
+    <Name>Strength</Name>
+  </Stat>
```
Would more elegant to have
`<Stat statType="Strength" name="Strength" />`

Grenkin1988 commented on Assets/Scripts/Models/Character.cs:
```diff
@@ -452,6 +493,29 @@ private void LoadNeeds()
+
+    public Stat GetStat(string statType)
+    {
+        Stat s = null;
```
`Stat stat = null;`

Grenkin1988 commented on Assets/Scripts/Models/Character.cs:
```diff
@@ -397,6 +410,27 @@ public void ReadXml(XmlReader reader)
+            foreach (string statString in statsStringArray)
+            {
+                string[] keyValue = statString.Split(';');
+                Stat s = GetStat(keyValue[0]);
```
`Stat stat`

Grenkin1988 commented on Assets/Scripts/Models/Character.cs:
```diff
@@ -416,7 +450,14 @@ public string GetDescription()
+            statText.Append("\n").Append(s.Value.Name).Append(": ").Append(s.Value.Value);
+        }
+
+        return "A human astronaut." + needText + statText;
```
Use `StringBuilder` for all statistics.

Grenkin1988 commented on Assets/Scripts/Models/Character.cs:
```diff
@@ -94,6 +98,7 @@ public Character(Tile tile, Color color)
         CurrTile = DestTile = nextTile = tile;
         characterColor = color;
         LoadNeeds();
+        LoadStats();
```
Extract methods call
```
LoadNeeds();
LoadStats();
```
to separate method.

mbstraus commented:
@Grenkin1988 

I think that commit should address your concerns above.

Grenkin1988 commented on Assets/Scripts/Models/Character.cs:
```diff
@@ -444,20 +447,19 @@ public string GetName()
         {
             // TODO: Localization
-            statText.Append("\n").Append(s.Value.Name).Append(": ").Append(s.Value.Value);
+            description.Append("\n").Append(s.Value.Name).Append(": ").Append(s.Value.Value);
```
-> description.AppendLine(string.Format("{0}: {1}", s.Value.Name, s.Value.Value));

Grenkin1988 commented on Assets/Scripts/Models/Character.cs:
```diff
@@ -444,20 +447,19 @@ public string GetName()
         }
 
-        StringBuilder statText = new StringBuilder();
         foreach (KeyValuePair<string, Stat> s in stats)
```
If you need only `Values` you should iterate by `stats.Values`

TomMalbran commented on Assets/Scripts/Models/Character.cs:
```diff
@@ -357,6 +366,13 @@ public void WriteXml(XmlWriter writer)
+        {
+            statsString.Append(stat.Value.statType).Append(";").Append(stat.Value.Value).Append(":");
+        }
+        writer.WriteAttributeString("stats", statsString.ToString());
```
I know that it might be easier to make your own parser. But is better to follow the xml way and save the stats like:
```
<Stats>
  <Stat name="Strength" value="10" />
</Stats>
```
It will be easier to extend and read when we add more attributes to the stats like xp.

mbstraus replied:
I was following the same style that the Needs were using, thus why I went with that way.  I do agree that it is cleaner to save them as separate elements like that.. I can certainly make that change... but unless it is desired, i would prefer not to touch needs unless I have to, but it will be inconsistent if I don't.

TomMalbran replied:
Oh. I didn't saw that the needs where done in that way. The needs code has several issues that need fixing. You shouldn't change the Needs in this PR since this PR is about Stats, but we should fix them in another PR.

TomMalbran commented:
You should also fix the StyleCop errors, I can see a few by just reading the code.

mbstraus commented:
I think that should take care of things.  You guys are tough, I like it :)

vogonistic commented:
This branch has merge conflicts. Please remove [conflicts] from the title when fixed.

mbstraus commented:
Conflicts have been resolved.

vogonistic commented:
Merged!

vogonistic commented:
@mbstraus I missed the StyleCop errors in this PR. Can you fix them and submit a new PR?

mbstraus commented:
Sure, although I won't be able to till tomorrow.

mbstraus commented:
@vogonistic Looks like #1083 is being worked to take care of all of the StyleCop issues.
